contract/src/Custody.sol

```diff
@@ -27,459 +28,276 @@ contract Custody is IChannel, IDeposit {
     error InvalidAmount();
     error TransferFailed(address token, address to, uint256 amount);
     error ChallengeNotExpired();
-    error ChannelNotFinal();
     error InsufficientBalance(uint256 available, uint256 required);
 
-    // Index in the array of participants
-    uint256 constant HOST = 0;
-    uint256 constant GUEST = 1;
-
     // Recommended structure to keep track of states
     struct Metadata {
-        Channel chan; // Opener define channel configuration
-        Status status; // Current channel status
-        uint256 challengeExpire; // If non-zero channel will resolve to lastValidState when challenge Expires
-        State lastValidState; // Last valid state when adjudicator was called
-        mapping(address token => uint256 balance) tokenBalances; // Token balances for the channel
+      Channel chan; // Opener define channel configuration
+      Status stage;
+      address creator;
+      Amount[] expectedDeposits; // Creator defines Token per participant
+      Amount[] actualDeposits; // Tracks deposits made by each participant
+      uint256 challengeExpire; // If non-zero channel will resolve to lastValidState when challenge Expires
+      State lastValidState; // Last valid state when adjudicator was called
+      mapping(address token => uint256 balance) tokenBalances; // Token balances for the channel
     }
 
     // Account is a ledger account per unique depositor and token
     struct Account {
-        uint256 available; // Available amount that can be withdrawn or allocated to channels
-        uint256 locked; // Amount currently allocated to channels
+      uint256 available; // Available amount that can be withdrawn or allocated to channels
+      uint256 locked; // Amount currently allocated to channels
     }
 
     struct Ledger {
-        mapping(address token => Account funds) tokens; // Token balances
-        EnumerableSet.Bytes32Set channels; // Set of user ChannelId
+      mapping(address token => Account funds) tokens; // Token balances
+      EnumerableSet.Bytes32Set channels; // Set of user ChannelId
     }
 
     mapping(bytes32 channelId => Metadata chMeta) internal _channels;
     mapping(address account => Ledger ledger) internal _ledgers;
 
     function deposit(address token, uint256 amount) external payable {
-        address account = msg.sender;
-        if (token == address(0)) {
-            require(msg.value == amount, InvalidAmount());
-        } else {
-            bool success = IERC20(token).transferFrom(account, address(this), amount);
-            require(success, TransferFailed(token, address(this), amount));
-        }
-        _ledgers[msg.sender].tokens[token].available += amount;
+      address account = msg.sender;
+      if (token == address(0)) {
+        require(msg.value == amount, InvalidAmount());
+      } else {
+        //TODO: Support native token
```

Comment:
It is already supported by the line above and `payable` modifier of this method.